File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/service/NotificationService.java
package com.github.arhor.simple.expense.tracker.service;

import java.io.IOException;

import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import com.github.arhor.simple.expense.tracker.model.NotificationDTO;

public interface NotificationService {

    SseEmitter subscribe(Long subscriberId);

    void unsubscribe(Long subscriberId);

    void handleNotification(Long senderId, Long userId, NotificationDTO notification);

    void sendNotification(Long senderId, Long userId, NotificationDTO notification) throws IOException;

    void sendNotifications() throws Exception;
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/service/NotificationService.kt
package com.github.arhor.simple.expense.tracker.service

import com.github.arhor.simple.expense.tracker.model.NotificationDTO
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter

interface NotificationService {

    fun subscribe(subscriberId: Long): SseEmitter

    fun unsubscribe(subscriberId: Long)

    fun unsubscribeAll()

    fun handleNotification(sourceUserId: Long, targetUserId: Long, notification: NotificationDTO)

    fun sendNotification(sourceUserId: Long, targetUserId: Long, notification: NotificationDTO)

    fun sendNotifications()
}
